test/test_mcmc.py

```diff
@@ -0,0 +1,62 @@
+from numpy.testing import assert_allclose
+
+import jax.numpy as np
+import jax.random as random
+from jax import jit, lax
+from jax.scipy.special import expit
+
+import numpyro.distributions as dist
+from numpyro.distributions.util import validation_disabled
+from numpyro.mcmc import hmc_kernel
+from numpyro.util import scan
+
+
+def test_unnormalized_normal():
+    true_mean, true_std = 1., 2.
+    warmup_steps, num_samples = 1000, 8000
+
+    def potential_fn(z):
+        return 0.5 * np.sum(((z[0] - true_mean) / true_std) ** 2)
+
+    def kinetic_fn(r, m_inv):
+        return 0.5 * np.sum(m_inv * r[0] ** 2)
+
+    init_kernel, sample_kernel = hmc_kernel(potential_fn, kinetic_fn)
+    init_samples = [np.array([0.])]
+    hmc_state = init_kernel(init_samples,
+                            num_warmup_steps=warmup_steps)
+    sample_kernel = jit(sample_kernel)
+    hmc_states = lax.scan(sample_kernel, hmc_state, np.arange(num_samples))
+    zs = hmc_states.z[0]
+    assert_allclose(np.mean(zs), true_mean, rtol=0.05)
+    assert_allclose(np.std(zs), true_std, rtol=0.05)
+
+
+def test_logistic_regression():
+    N, dim = 3000, 3
+    warmup_steps, num_samples = 1000, 8000
+    data = random.normal(random.PRNGKey(0), (N, dim))
+    true_coefs = np.arange(1., dim + 1.)
+    probs = expit(np.sum(true_coefs * data, axis=-1))
+    labels = dist.bernoulli(probs).rvs()
+
+    with validation_disabled():
+        def potential_fn(beta):
+            coefs_mean = np.zeros(dim)
+            coefs_lpdf = dist.norm(coefs_mean, np.ones(dim)).logpdf(beta)
+            probs = expit(np.sum(beta * data, axis=-1))
+            y_lpdf = dist.bernoulli(probs).logpmf(labels)
+            return - (np.sum(coefs_lpdf) + np.sum(y_lpdf))
+
+        def kinetic_fn(beta, m_inv):
+            return 0.5 * np.dot(m_inv, beta ** 2)
+
+        init_kernel, sample_kernel = hmc_kernel(potential_fn, kinetic_fn)
+        init_samples = np.zeros(dim)
+        hmc_state = init_kernel(init_samples,
+                                step_size=0.01,
```

Comment:
One thing that I noticed is that HMC in numpyro is way more sensitive to these hyper-parameters and will give a `nan` or `inf` values easily if these parameters are too high.

Reply:
The `nan`, `inf` are natural values for this model if step_size is large I guess. Might be find_reasonable_step_size does not work as expected. I'll take a look too.

Reply:
neerajprad There are a few things happening here:
+ Starting with step_size=0.01, num_steps=15 is equivalent to set trajectory_length=0.15. If we set step_size=1, num_steps=15, then each trajectories will require a large number of steps. So the inference will be slower. Anyway, this is not the reason why we get `nan`, `inf` so often.
+ I think that the reason is: pytorch clamps parameters while numpyro did not. For example,
`torch_dist.Bernoulli(1).log_prob(0) = -16` while `jax_dist.bernoulli(1).logpmf(0) = -inf`. Of course, -inf is the correct value but having -inf log_prob will easily lead to grad of latent is +-inf or nan.

To fix this issue, I suggest to clamp `probs` in potential_fn:
```
            probs = np.clip(expit(np.sum(beta * data, axis=-1)),
                            a_min=np.finfo(np.float32).tiny,
                            a_max=(1 - np.finfo(np.float32).eps))
```

Later, we should do this automatically for users.

Reply:
Thanks a lot, @fehiepsi for digging into this! I have followed this suggestion and also adjusted the step size to higher than what it was previously.